// structurae@4.0.0/esm/adjacency-matrix-weighted-undirected.js
import { AdjacencyMatrixWeightedDirectedMixin } from "./adjacency-matrix-weighted-directed.js";
/**
 * Creates an Adjacency Matrix class extending a given TypedArray class.
 *
 * @param Base a TypedArray class to extend
 */
export function AdjacencyMatrixWeightedUndirectedMixin(Base) {
    /**
     * Implements the Adjacency Matrix for weighted undirected graphs.
     */
    class AdjacencyMatrixWeightedUndirected extends AdjacencyMatrixWeightedDirectedMixin(Base) {
        static getLength(vertices, _) {
            return ((vertices + 1) * vertices) >> 1;
        }
        static getVertices(length) {
            return (Math.sqrt((length << 3) + 1) - 1) >> 1;
        }
        getIndex(x, y) {
            return x >= y ? y + (((x + 1) * x) >> 1) : x + (((y + 1) * y) >> 1);
        }
        outEdges(vertex) {
            // TODO: fix type casting
            return this.inEdges(vertex);
        }
    }
    Object.defineProperty(AdjacencyMatrixWeightedUndirected, "directed", {
        enumerable: true,
        configurable: true,
        writable: true,
        value: false
    });
    // TODO: fix type casting
    return AdjacencyMatrixWeightedUndirected;
}
